File: src/cameo/TimeCondition.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "TimeCondition.h"
#include <boost/date_time/posix_time/posix_time.hpp>
#include <iostream>

using namespace std;

namespace cameo {
	
TimeCondition::TimeCondition() :
	m_notified(false) {
}

TimeCondition::~TimeCondition() {
}
	
bool TimeCondition::wait(long timeMs) {

	boost::mutex::scoped_lock lock(m_conditionMutex);
	
	if (!m_notified) {
		boost::posix_time::time_duration duration = boost::posix_time::millisec(timeMs);

		return m_condition.timed_wait(lock, duration);
	}
	
	// sure that notify occurred
	return true;
}

void TimeCondition::notify() {

	boost::mutex::scoped_lock lock(m_conditionMutex);

	m_notified = true;	
	m_condition.notify_one();
}

}

File: java/api/src/main/java/fr/ill/ics/cameo/base/TimeCondition.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.base;

/**
 * Class defining a time condition.
 */
public class TimeCondition {

	private boolean notified = false;

	/**
	 * Waits for the time.
	 * @param timeMs Time in milliseconds.
	 * @return True if notification occurred during the wait.
	 */
	public synchronized boolean waitFor(int timeMs) {
		
		try {
			super.wait(timeMs);
		}
		catch (InterruptedException e) {
			return true;
		}
		
		return notified;
	}
	
	/**
	 * Notifies the condition. It unblocks the waitFor() call.
	 */
	public synchronized void notifyCondition() {
		notified = true;
		notify();
	}
}
